src/binding/c/topo_api.txt

```diff
@@ -0,0 +1,154 @@
+    .skip: global_cs
+    .extra: ignore_revoked_comm
+{ -- error_check -- coords
+    if (maxdims > 0) {
+        MPIR_ERRTEST_ARGNULL(coords, "coords", mpi_errno);
+    }
+}
+
+MPI_Cart_create:
+    .desc: Makes a new communicator to which topology information has been attached
+    .extra: errtest_comm_intra
+    .impl: topo_fns->cartCreate
+
+MPI_Cart_get:
+    .desc: Retrieves Cartesian topology information associated with a communicator
+    .extra: ignore_revoked_comm
+{ -- error_check -- dims, periods, coords
+    if (maxdims > 0) {
+        MPIR_ERRTEST_ARGNULL(dims, "dims", mpi_errno);
+        MPIR_ERRTEST_ARGNULL(periods, "periods", mpi_errno);
+        MPIR_ERRTEST_ARGNULL(coords, "coords", mpi_errno);
+    }
+}
+
+MPI_Cart_map:
+    .desc: Maps process to Cartesian topology information
+    .skip: global_cs
+    .extra: ignore_revoked_comm
+    .impl: topo_fns->cartMap
+
+MPI_Cart_rank:
+    .desc: Determines process rank in communicator given Cartesian location
+    .skip: global_cs
+    .extra: ignore_revoked_comm
+/*
+    Notes:
+     Out-of-range coordinates are erroneous for non-periodic dimensions.
+     Versions of MPICH before 1.2.2 returned 'MPI_PROC_NULL' for the rank in this
+     case.
+*/
+{ -- error_check --
+    MPIR_Topology *cart_ptr = MPIR_Topology_get(comm_ptr);
+    MPIR_ERR_CHKANDJUMP((!cart_ptr || cart_ptr->kind != MPI_CART),
+                        mpi_errno, MPI_ERR_TOPOLOGY, "**notcarttopo");
+    int ndims = cart_ptr->topo.cart.ndims;
+    if (ndims != 0) {
+        MPIR_ERRTEST_ARGNULL(coords, "coords", mpi_errno);
+    }
+    for (int i = 0; i < ndims; i++) {
+        if (!cart_ptr->topo.cart.periodic[i]) {
+            int coord = coords[i];
+            MPIR_ERR_CHKANDJUMP3((coord < 0 || coord >= cart_ptr->topo.cart.dims[i]),
+                                  mpi_errno, MPI_ERR_ARG, "**cartcoordinvalid",
+                                  "**cartcoordinvalid %d %d %d", i, coords[i],
+                                  cart_ptr->topo.cart.dims[i] - 1);
+        }
+    }
+}
+
+MPI_Cart_shift:
+    .desc: Returns the shifted source and destination ranks, given a shift direction and amount
+    .skip: global_cs
+    .extra: ignore_revoked_comm
+/*
+    Notes:
+    The 'direction' argument is in the range '[0,n-1]' for an n-dimensional
+    Cartesian mesh.
+*/
+{ -- error_check -- disp
+}
+
+MPI_Cart_sub:
+    .desc: Partitions a communicator into subgroups which form lower-dimensional cartesian subgrids
+
+MPI_Cartdim_get:
+    .desc: Retrieves Cartesian topology information associated with a communicator
+    .skip: global_cs
+    .extra: ignore_revoked_comm
+
+MPI_Dims_create:
+    .desc: Creates a division of processors in a cartesian grid
+
+MPI_Dist_graph_create:
+    .desc: MPI_DIST_GRAPH_CREATE returns a handle to a new communicator to which the distributed graph topology information is attached
+    .extra: errtest_comm_intra
```

Comment:
Is it okay to remove the following check?
```c
            if (n > 0) {
                int have_degrees = 0;
                MPIR_ERRTEST_ARGNULL(sources, "sources", mpi_errno);
                MPIR_ERRTEST_ARGNULL(degrees, "degrees", mpi_errno);
                for (i = 0; i < n; ++i) {
                    if (degrees[i]) {
                        have_degrees = 1;
                        break;
                    }
                }
                if (have_degrees) {
                    MPIR_ERRTEST_ARGNULL(destinations, "destinations", mpi_errno);
                    if (weights != MPI_UNWEIGHTED)
                        MPIR_ERRTEST_ARGNULL(weights, "weights", mpi_errno);
                }
            }
```

Reply:
Added